### ISSUE TITLE
libcrypto-1_1.dll and libssl-1_1.dll compile with non-ideal linker flags on Windows

### ISSUE THREAD

**Issue by dmachaj** [openssl/openssl#18933]
# Bug report

I am expanding [BinSkim](https://github.com/microsoft/binskim) to check for some performance options when analyzing Windows binaries.  I am using Python as part of the test matrix and discovered that two of the binaries that ship with Python do not have ideal linker flags.  (This is still a work in progress, so it requires a work-in-progress build of BinSkim.)  

These files are not part of the Python code base so this issue is redirecting [https://github.com/python/cpython/issues/95551](https://github.com/python/cpython/issues/95551) to the correct project.

The output that I am seeing is as follows:
```
v:\T\binskim\bld\bin\x64_Debug\netcoreapp3.1\BinSkim.exe analyze V:\OSS-Perf\Python\python-3.10.4\DLLs\libcrypto-1_1.dll --verbose
V:\OSS-Perf\Python\python-3.10.4\DLLs\libcrypto-1_1.dll: warning BA6001: 'libcrypto-1_1.dll' appears to be compiled as release but enables incremental linking, increasing binary size and further compromising runtime performance by preventing enabling maximal code optimization.
V:\OSS-Perf\Python\python-3.10.4\DLLs\libcrypto-1_1.dll: warning BA6004: 'libcrypto-1_1.dll' was compiled with COMDAT folding (/OPT:ICF) disabled, increasing binary size.
```

```
v:\T\binskim\bld\bin\x64_Debug\netcoreapp3.1\BinSkim.exe analyze V:\OSS-Perf\Python\python-3.10.4\DLLs\libssl-1_1.dll --verbose
V:\OSS-Perf\Python\python-3.10.4\DLLs\libssl-1_1.dll: warning BA6001: 'libssl-1_1.dll' appears to be compiled as release but enables incremental linking, increasing binary size and further compromising runtime performance by preventing enabling maximal code optimization.
V:\OSS-Perf\Python\python-3.10.4\DLLs\libssl-1_1.dll: warning BA6004: 'libssl-1_1.dll' was compiled with COMDAT folding (/OPT:ICF) disabled, increasing binary size.
```

See also:
[https://docs.microsoft.com/en-us/cpp/build/reference/opt-optimizations?view=msvc-170#arguments](https://docs.microsoft.com/en-us/cpp/build/reference/opt-optimizations?view=msvc-170#arguments)
[https://docs.microsoft.com/en-us/cpp/build/reference/incremental-link-incrementally?view=msvc-170](https://docs.microsoft.com/en-us/cpp/build/reference/incremental-link-incrementally?view=msvc-170)

# Your environment

This analysis was performed on a download of Python 3.10.4.  That was the current release the last time I downloaded it.

The relevant rule additions for BinSkim are:
[BA6004.EnableCOMDATFolding](https://github.com/microsoft/binskim/issues/679)
[BA6001.DisableIncrementalLinkingInReleaseBuilds](https://github.com/microsoft/binskim/pull/667)

**Comment by mattcaswell** [comment]
Reading those two MS links I read that:

> By default, the linker runs in incremental mode.

And that:

> Most programs can be linked incrementally. However, some changes are too great, and some options are incompatible with incremental linking. LINK performs a full link if any of the following options are specified:
> ...snip...
>/OPT:ICF is selected

And that:

> By default, /OPT:ICF is enabled by the linker unless /OPT:NOICF or /DEBUG is specified.

So, if I read that correctly, `/DEBUG` will turn off `/OPT:ICF` which also means incremental linking will be used (unless some other option prevents it). But production builds (no `/DEBUG`) should have `/OPT:ICF` on, and full linking should occur (and COMDAT folding should be on). I don't see anywhere that we directly set or unset the ICF option.

I also see this in 10-main.conf for the `VC-common` target (which is a common ancestor for all `VC-*` windows targets in an OpenSSL build):

https://github.com/openssl/openssl/blob/4842a27b902660b672d72d2ed23e941461ca481c/Configurations/10-main.conf#L1335-L1343

Note the `/debug` entry for `LDFLAGS`. This looks suspicious. I don't see anything in there that suppresses `/debug` for production builds, but maybe I'm missing it (very possible).

Probably it needs someone more Windowsy than I am to investigate this further. I'd ask @levitte but he is on vacation. Setting "Help wanted" for now.

**Comment by dmachaj** [comment]
(My changes to BinSkim are still in testing.  I fixed a bug and it produced an additional warning that should have been in the original bug report.  I added it up above.)

`/debug` is a good thing to pass because it means "produce a PDB file so that debugging is possible".  Debugging should be possible on release builds too :).

Assuming `LDFLAGS` is the set of options passed to the linker then adding `/OPT:REF,ICF` and `/INCREMENTAL:NO` for release builds would probably be ideal.  Incremental link is fine for non-release builds.  `/OPT:ICF` can make debugging harder so it is not ideal for debug builds.

In short:

## Release
`/nologo /debug /OPT:REF,ICF /INCREMENTAL:NO`

## Debug
`/nologo /debug /OPT:REF,NOICF /INCREMENTAL`

**Comment by dmachaj** [comment]
I am doing some additional testing of my BinSkim change so I had the linker command-line for these binaries handy.  Curiously some of the arguments are duplicated.  I don't think that harms anything, but it is a bit odd.

`"link.exe /nologo /debug /dll /nologo /debug /implib:libssl.lib /out:libssl-1_1.dll /def:libssl.def"`

**Comment by mattcaswell** [comment]
@levitte - do you have any input?

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 2 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity for more than 2 months.
